Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp

```diff
@@ -90,4 +93,62 @@ void GraphicsLayerCARemote::moveToContext(RemoteLayerTreeContext& context)
     context.graphicsLayerDidEnterContext(*this);
 }
 
+class GraphicsLayerCARemoteAsyncContentsDisplayDelegate : public GraphicsLayerAsyncContentsDisplayDelegate {
+public:
+    GraphicsLayerCARemoteAsyncContentsDisplayDelegate(IPC::Connection& connection, DrawingAreaIdentifier identifier, WebCore::GraphicsLayer::PlatformLayerID layerID)
+        : m_connection(connection)
+        , m_drawingArea(identifier)
+        , m_layerID(layerID)
+    { }
+
+    bool hasShareableBackend(ImageBuffer& buffer)
+    {
+        auto* backend = buffer.ensureBackendCreated();
+        if (!backend)
+            return false;
+
+        buffer.flushDrawingContext();
+
+        auto* sharing = backend->toBackendSharing();
+        return is<ImageBufferBackendHandleSharing>(sharing);
+    }
+
+    bool tryCopyToLayer(ImageBuffer& buffer) final
+    {
+        if (!hasShareableBackend(buffer))
+            return false;
+        auto clone = buffer.clone();
+        if (!clone)
+            return false;
+        ASSERT(hasShareableBackend(*clone));
+        auto* backend = clone->ensureBackendCreated();
+        if (!backend)
+            return false;
+
+        clone->flushDrawingContext();
```

Comment:
The ASSERT above had the side-effect of flushing, but you flush here again. Side effects are bad!